pkg/install/stack/kubermatic-master/certmanager.go

```diff
@@ -18,21 +18,374 @@ package kubermaticmaster
+
+	// step 1: purge the Helm release
+	logger.Info("Uninstalling release…")
+	if err := helmClient.UninstallRelease(CertManagerNamespace, CertManagerReleaseName); err != nil {
+		return fmt.Errorf("failed to uninstall release: %v", err)
+	}
+
+	now := time.Now().Format("2006-01-02T150405")
+
+	// for these CRDs, we not only back them up as YAML, but also restore them
+	// automatically so the user doesn't have to (and we filter out resources
+	// with an ownerRef, which would indicate that some other process/object
+	// manages a given certificate, for example)
+	restorableCRDs := []schema.GroupVersionKind{
+		// restore issuers and clusterissues before certs and requests
+		{Version: "v1alpha2", Group: "cert-manager.io", Kind: "clusterissuer"},
+		{Version: "v1alpha2", Group: "cert-manager.io", Kind: "issuer"},
+		{Version: "v1alpha2", Group: "cert-manager.io", Kind: "certificaterequest"},
+		{Version: "v1alpha2", Group: "cert-manager.io", Kind: "certificate"},
+	}
+
+	allCRDs := append(
+		restorableCRDs,
+		schema.GroupVersionKind{Version: "v1alpha2", Group: "acme.cert-manager.io", Kind: "challenge"},
+		schema.GroupVersionKind{Version: "v1alpha2", Group: "acme.cert-manager.io", Kind: "order"},
+	)
+
+	// step 2: fetch restorable resources in memory
+	logger.Info("Creating backups for all Custom Resources…")
+	objectsToRestore, secrets, err := getCustomResources(ctx, logger, kubeClient, restorableCRDs)
+	if err != nil {
+		return fmt.Errorf("failed to list resources: %v", err)
+	}
+
+	// step 3: backup resources into files
+	for _, crdGVK := range allCRDs {
+		logger.Infof("  dumping %s", crdGVK.Kind)
+
+		filename := fmt.Sprintf("backup_%s_%s.yaml", now, crdGVK.Kind)
+		if err := util.BackupResources(ctx, kubeClient, crdGVK, filename); err != nil {
+			return fmt.Errorf("failed to backup %s resources: %v", crdGVK.Kind, err)
+		}
+	}
+
+	logger.Infof("  dumping secret")
+	filename := fmt.Sprintf("backup_%s_secret.yaml", now)
+	if err := util.DumpResources(ctx, filename, secrets); err != nil {
+		return fmt.Errorf("failed to backup secret resources: %v", err)
+	}
+
+	// step 4: remove finalizers from resources
+	logger.Info("Removing finalizers from Custom Resources…")
+	if err := removeFinalizersFromCustomResources(ctx, kubeClient, allCRDs, []string{"finalizer.acme.cert-manager.io"}); err != nil {
+		return fmt.Errorf("failed to remove finalizers: %v", err)
+	}
+
+	// step 5: delete all cert-manager CRDs
+	logger.Info("Removing Custom Resource Definitions…")
+	for _, crdGVK := range allCRDs {
+		crd := apiextensionsv1.CustomResourceDefinition{}
+		key := types.NamespacedName{Name: crdName(crdGVK)}
+
+		if err := kubeClient.Get(ctx, key, &crd); err != nil {
+			if kerrors.IsNotFound(err) {
+				continue
+			}
+
+			return fmt.Errorf("failed to retrieve CRD %s: %v", crdGVK.Kind, err)
+		}
+
+		if err := kubeClient.Delete(ctx, &crd); err != nil {
+			return fmt.Errorf("failed to delete CRD %s: %v", crdGVK.Kind, err)
+		}
+	}
```

Comment:
Dont think you need GET and then DELETE, this should be simpler:

```suggestion
	for _, crdGVK := range allCRDs {
		crd := apiextensionsv1.CustomResourceDefinition{}
		crd.Name = crdName(crdGVK)

		if err := kubeClient.Delete(ctx, &crd); err != nil {
			if kerrors.IsNotFound(err) {
				continue
			}

			return fmt.Errorf("failed to delete CRD %s: %v", crdGVK.Kind, err)
		}
	}
```

Reply:
Also wont this delete also the `v1beta1` and `v1` certmanager CRD's if the user has them? Maybe we should do a version check before the deletion, or

```
crd  := apiextensionv1beta1.CustomResourceDefinition{}

``` 
instead of the 
```
crd = apiextensionsv1.CustomResourceDefinition{}
```

Not sure if that will work

Reply:
> Also wont this delete also the v1beta1 and v1 certmanager CRD's if the user has them? Maybe we should do a version check before the deletion, or

>crd  := apiextensionv1beta1.CustomResourceDefinition{}
>instead of the

>crd = apiextensionsv1.CustomResourceDefinition{}
>Not sure if that will work

Discussed internally. In the case that the migration malfunctions because of `v1beta1` apiVersion of the CRDs, the user needs to do a manual migration, using the backups. As 2.16 and older Kubermatic ships with the `v1alpha2` version and 2.17 and newer one with the `v1` version of cert-manager CRD, the in-between state(`v1beta`) wont have automatic migration support as its uncharted teritory.